### PR TITLE
🚨🚨 Source SFTP Bulk: migrate to file-based CDK

### PR DESCRIPTION
**Pull request by artem1205**
## What

Resolve https://github.com/airbytehq/airbyte-internal-issues/issues/6666

## How

migrate to File-Based CDK

## Recommended reading order
1. `*.py`

## 🚨 User Impact 🚨

- breaking changes in spec (auth selector)
- migration to file-based CDK

## Pre-merge Actions

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            )
+        return self._sftp_client
+
+    def get_matching_files(
+        self,
+        globs: List[str],
+        prefix: Optional[str],
+        logger: logging.Logger,
+    ) -> Iterable[RemoteFile]:
+        # TODO: check prefixes
```
Comment (tolik0) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:60]: Should these lines be deleted or uncommented?

File: airbyte-integrations/connectors/source-sftp-bulk/integration_tests/files/csv/test_3.csv [hunk 2]
```diff
@@ -0,0 +1,3 @@
+string_col,int_col
+"hello",1
+"foo",2
```
Comment (askarpets) [inline airbyte-integrations/connectors/source-sftp-bulk/integration_tests/files/csv/test_3.csv:3]: Please add an empty line

File: airbyte-integrations/connectors/source-sftp-bulk/integration_tests/utils.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+import json
+import logging
+import os
+import re
+from typing import Any, Mapping, Union
+
+logger = logging.getLogger("airbyte")
+
+PRIVATE_KEY = str
```
Comment (askarpets) [inline airbyte-integrations/connectors/source-sftp-bulk/integration_tests/utils.py:12]: Shouldn't an empty string be used here as a default value?

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py [hunk 4]
```diff
@@ -2,34 +2,29 @@
...
+def handle_backoff(details):
+    logger.warning("SSH Connection closed unexpectedly. Waiting {wait} seconds and retrying...".format(**details))
 
 
 class SFTPClient:
-    _connection = None
+    _connection: paramiko.SFTPClient = None
 
-    def __init__(self, host, username, password=None, private_key=None, port=None, timeout=REQUEST_TIMEOUT):
+    def __init__(self, host: str, username: str, password: str = None, private_key=None, port: int = None, timeout=REQUEST_TIMEOUT):
```
Comment (askarpets) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py:27]: Please use `Optional[str]` if default is present and add annotation for the `timeout` param

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py (hunk 1 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py]: deleted

**Review by askarpets — APPROVED** [review]

File: airbyte-integrations/connectors/source-sftp-bulk/integration_tests/utils.py (hunk 3 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-sftp-bulk/integration_tests/utils.py:12]: agreed

**Review by girarda — APPROVED** [review]
question on whether this needs to be a breaking change, but looks good otherwise!

File: docs/integrations/sources/sftp-bulk-migrations.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
+# SFTP Bulk Migration Guide
```
[the remark this answers is not in the document]
Comment (artem1205) [reply docs/integrations/sources/sftp-bulk-migrations.md:1]: @girarda ,
I believe we can migrate the configuration, but introduced changes may still cause disruptions due to the nature of file-based CDK, including additional fields and incremental sync.
Additionally, I have decided to introduce breaking changes due to the _community_ status of this connector and the presence of only two active cloud connections.

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    ) -> Iterable[RemoteFile]:
+        directories = [self._config.folder_path or "/"]
+
+        # Iterate through directories and subdirectories
+        while directories:
+            current_dir = directories.pop()
+            try:
+                items = self.sftp_client.sftp_connection.listdir_attr(current_dir)
+            except FileNotFoundError:
+                continue
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:68]: added

File: airbyte-integrations/connectors/source-sftp-bulk/metadata.yaml [hunk 7]
```diff
@@ -22,8 +24,13 @@ data:
     oss:
       enabled: true
   releaseStage: alpha
+  releases:
+    breakingChanges:
+      1.0.0:
+        message: "This upgrade brings migration to File-Based CDK"
+        upgradeDeadline: "2024-04-08"
```
Comment (katmarkham) [inline airbyte-integrations/connectors/source-sftp-bulk/metadata.yaml:31]: Is this meant to be immediate or should we switch the date to two weeks in the future?

File: airbyte-integrations/connectors/source-sftp-bulk/metadata.yaml [hunk 8]
```diff
@@ -22,8 +24,13 @@ data:
     oss:
       enabled: true
   releaseStage: alpha
+  releases:
+    breakingChanges:
+      1.0.0:
+        message: "This upgrade brings migration to File-Based CDK"
```
Comment (katmarkham) [inline airbyte-integrations/connectors/source-sftp-bulk/metadata.yaml:30]: This upgrade migrates the SFTP Bulk source to the Airbyte file-based CDK.  This is the first necessary step of transitioning a file connector from community to Airbyte maintained.

File: docs/integrations/sources/sftp-bulk-migrations.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+## Upgrading to 1.0.0
+This release upgrades the SFTP Bulk connector to file-based CDK which causes the following changes:
+
+- configuration changes
+- stream changes
```
Comment (katmarkham) [inline docs/integrations/sources/sftp-bulk-migrations.md:7]: - Configuration changes
- Stream changes

**Review by katmarkham — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-sftp-bulk/metadata.yaml (hunk 7 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-sftp-bulk/metadata.yaml:31]: updated

File: airbyte-integrations/connectors/source-sftp-bulk/metadata.yaml (hunk 8 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-sftp-bulk/metadata.yaml:30]: updated

File: docs/integrations/sources/sftp-bulk-migrations.md (hunk 9 above)
Comment (artem1205) [reply docs/integrations/sources/sftp-bulk-migrations.md:7]: updated

**Review by katmarkham — APPROVED** [review]

**Pull request merged**